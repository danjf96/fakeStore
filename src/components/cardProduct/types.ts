import { ColorSchemeName, TextStyle, TouchableOpacityProps, ViewStyle } from "react-native";
import { PRODUCTS } from "../../store/ducks/products/types";

export interface PRODUCTCARDPROPS {
    styleContainer?: ViewStyle,
    data: CARDPRODUCCTPROPS,
    onPressButton: (type: 'add' | 'rm',data:CARDPRODUCCTPROPS) => void,
    theme?: ColorSchemeName
}

export interface CARDPRODUCCTPROPS extends PRODUCTS {
    "quantity": number
}

