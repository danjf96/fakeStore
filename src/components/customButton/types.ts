import { TextStyle, TouchableOpacityProps } from "react-native";

export interface BUTTONPROPS  extends TouchableOpacityProps {
    text?: string,
    stylesText?: TextStyle
}