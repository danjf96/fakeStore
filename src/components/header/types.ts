import { ColorSchemeName, ImageSourcePropType } from "react-native";

export interface PROPSHEADER {
    title?: string,
    buttonLeft?: ImageSourcePropType | string,
    buttonRight?: ImageSourcePropType,
    theme?: ColorSchemeName
    onPressLeft?: () => void,
    onPressRight?: () => void,
}