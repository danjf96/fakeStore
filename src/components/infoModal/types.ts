export interface INFOMODALPROPS {
    title: string,
    text: string,
    btnSuccess: () => void,
    textSuccess: string,
    btnCancel: () => void
    textCancel: string,
    visible: boolean
}