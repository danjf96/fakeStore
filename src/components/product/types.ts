import { ColorSchemeName, TextStyle, TouchableOpacityProps, ViewProps, ViewStyle } from "react-native";

export interface PRODUCTPROPS extends ViewProps {
    styleContainer?: ViewStyle,
    data: Product,
    typeStyle?: 'big' | 'small',
    onPressButton?: (data:Product) => void,
    theme?: ColorSchemeName
}

interface Product {
    "id": number,
    "title": string,
    "price": number,
    "description": string,
    "category":string,
    "image": string,
    "rating":{
        "rate" :number,
        "count": number
    }
}