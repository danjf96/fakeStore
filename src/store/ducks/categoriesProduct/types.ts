export interface STATECATEGORIESPRODUCT {
    loading: boolean,
    list: string[],
    category?: string
}

//TYPES ACTIONS
export const Types = {
    CLEAN: 'CLEAN',
    CLEAN_CATEGORIES: 'CLEAN_CATEGORIES',
    CHANGE: 'CHANGE_CATEGORIES',
    LOAD: 'LOAD_CATEGORIES'
}

