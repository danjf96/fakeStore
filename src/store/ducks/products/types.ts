export interface STATESTORE {
    loading: boolean
}

//TYPES ACTIONS
export const Types = {
    CLEAN: 'CLEAN',
    CHANGE: 'CHANGE_PRODUCTS',
    LOAD: 'LOAD_PRODUCTS'
}

