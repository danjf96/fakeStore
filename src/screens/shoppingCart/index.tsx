import React, { useState } from 'react'
import { FlatList, Image, Text, View } from 'react-native'
import Colors from '../../assets/Colors'
import { GlobalsStyles } from '../../assets/Globals'
import CardProduct from '../../components/cardProduct'
import { CARDPRODUCCTPROPS } from '../../components/cardProduct/types'
import Container from '../../components/container'
import CustomButton from '../../components/customButton'
import Header from '../../components/header'
import InfoModal from '../../components/infoModal'
import { useAppDispatch, useAppSelector } from '../../store'
import { changeCartProduct } from '../../store/ducks/shoppingCart'
import { formatMoney } from '../../utils/functions'
import Styles from './styles'
const left = require('../../assets/images/arrow_left.png')
const logo = require('../../assets/images/logo.png')

const ShoppingCart = (props:any) => {

    const dispatch = useAppDispatch()
    const { shoppingCart: { list,  total } } = useAppSelector( ({ shoppingCart  } ) => ({  shoppingCart }) )
    const [ openModal, setOpenModal ] = useState(false)
    const [ rmProduct, setRmProduct ] = useState<CARDPRODUCCTPROPS>()

    const finish = () => props.navigation.navigate('Finish')
    
    const changeValue = (type: 'add' | 'rm', product?: CARDPRODUCCTPROPS ) => {
        if(product?.quantity === 1 && type === 'rm' && !openModal){
            setRmProduct(product)
            setOpenModal(true)
        }
        else
            product && dispatch(changeCartProduct(type, list, product))
    }

    return (
        <>
            <Container style={{ height: '100%' }}>

                <Header buttonLeft={left} title='carrinho' onPressLeft={ () => props.navigation.goBack()}/>

                <Text style={{ ...GlobalsStyles.title, marginTop: 30 }}>Meu carrinho</Text>
                
                {list.length === 0 
                    ?
                    <View style={{ flexDirection: 'column', alignItems: 'center', justifyContent: 'center', marginTop: '50%'}}>
                        <Image source={logo}  style={Styles.imageEmpty}/>
                        <Text style={Styles.textEmpty}>Nenhum item adicionado no carrinho.</Text>
                        <CustomButton 
                            text='ADICIONAR ITENS'
                            onPress={() => props.navigation.navigate('Home')}
                            style={{ backgroundColor: '#8775FE'}}
                        />
                    </View>
                    
                    :
                    <>
                        <FlatList 
                            renderItem={ ({ item, index}) => <CardProduct 
                                    data={item}
                                    styleContainer={Styles.productContainer}
                                    onPressButton={changeValue}
                                    testID='cart'
                                />
                            }
                            keyExtractor={ (item,index) => `cart${index}`}
                            data={list}
                            style={{ marginVertical: 20, }}
                        />   

                    
                        <View style={Styles.containerPrice}>
                            <Text style={Styles.price}> Total: </Text>
                            <Text style={Styles.price}>R$ {formatMoney(total)}</Text>
                        </View>
                        
                            
                        <View style={{ width: '100%', paddingHorizontal: 29, paddingTop: 20  }}>
                            <CustomButton 
                                text='FINALIZAR COMPRA'
                                onPress={finish}
                            />
                        </View>
                    </>
                }   


                <InfoModal  
                    // testID='finishModal'
                    visible={openModal}
                    title={'Remover Item'} 
                    text={'Se deseja remover o item do carrinho clique em prosseguir.'} 
                    textSuccess={'PROSSEGUIR'} 
                    textCancel='CANCELAR' 
                    btnSuccess={() => {
                        setOpenModal(false)
                        changeValue('rm', rmProduct)
                    }} 
                    btnCancel={() => setOpenModal(false)}
                />  

            </Container>
        </>

    )
}

export default ShoppingCart