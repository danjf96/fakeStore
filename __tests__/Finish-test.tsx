/**
 * @format
 */

 import 'react-native';
 import React from 'react';
 import Finish from '../src/screens/finish';
 import { render, waitFor } from 'react-native-testing-library';
import { Provider } from 'react-redux';
import store from '../src/store';
import { ProductCart, Types } from '../src/store/ducks/shoppingCart/types';
 const component = (
  <Provider store={store} > <Finish navigation={{ navigate: (screen:'string') => null }}/> </Provider>
);

const fakeValues:ProductCart[] = [{
  "id":1,
  "title":"Title",
  "price":100,
  "description":"description",
  "category":"category",
  "image":"https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
  "rating":{"rate":3.9,"count":120},
  "quantity": 1
}]

describe('clean cart', () => {
  it('renders correctly', async () => {
    const screen = render(component)
    expect(screen.toJSON()).toMatchSnapshot()
  });

  store.dispatch({ type: Types.CHANGE, payload: { list: fakeValues, total: 10 }})
  const cart = store.getState().shoppingCart

  it('values', () => {
    expect(cart.list.length).not.toBe(0)
    expect(cart.total).toBe(10)
  
  })

  it('clean values', async () => {
    const screen = await waitFor( () => {
      return render(component)
    })
  
  
    await waitFor( () => {
      const { list, total } = store.getState().shoppingCart
  
      expect(list.length).toBe(0)
      expect(total).toBe(0)
    })  
  
  })

});
  