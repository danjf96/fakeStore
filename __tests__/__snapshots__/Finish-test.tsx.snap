// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clean cart renders correctly 1`] = `
Array [
  " ",
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#8775FE",
        "flexDirection": "column",
        "height": "100%",
        "justifyContent": "center",
        "width": "100%",
      }
    }
  >
    <Image
      collapsable={false}
      forwardedRef={[Function]}
      resizeMode="contain"
      source={
        Object {
          "testUri": "../../../src/assets/images/success.png",
        }
      }
      style={
        Object {
          "height": 150,
          "marginBottom": 30,
          "resizeMode": "contain",
          "tintColor": "white",
          "width": 150,
        }
      }
    />
    <Text
      style={
        Object {
          "color": "white",
          "fontFamily": "WorkSans-Regular",
          "fontSize": 25,
          "fontWeight": "700",
          "marginBottom": 12,
        }
      }
    >
      SUCESSO!
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "white",
            "fontFamily": "WorkSans-Regular",
            "fontSize": 25,
            "fontWeight": "700",
            "marginBottom": 12,
          },
          Object {
            "fontSize": 16,
            "fontWeight": "normal",
          },
        ]
      }
    >
      Compra realizada com sucesso, aproveite!
    </Text>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#504DB6",
          "borderRadius": 37,
          "bottom": 60,
          "flexDirection": "column",
          "height": 50,
          "justifyContent": "center",
          "opacity": 1,
          "position": "absolute",
          "width": "90%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "white",
              "fontFamily": "WorkSans-Regular",
              "fontSize": 14,
              "fontWeight": "700",
            },
            undefined,
          ]
        }
      >
        PROSSEGUIR
      </Text>
    </View>
  </View>,
  " ",
]
`;
